src/Servant/Multipart.hs

```diff
@@ -244,49 +244,32 @@ instance ( FromMultipart a
 -- returning the data as well as the resourcet InternalState
 -- that allows us to properly clean up the temporary files
 -- later on.
-check :: MultipartOptions -> DelayedIO (MultipartData, InternalState)
-check opts = withRequest $ \request -> liftIO $ do
-  st <- createInternalState
-  rawData <- parseRequestBodyEx parseOpts (tmpBackend opts st) request
-  return (fromRaw rawData, st)
-
+check :: MultipartOptions -> DelayedIO MultipartData
+check opts = withRequest $ \request -> do
+  st <- liftResourceT getInternalState
+  rawData <- liftIO $ parseRequestBodyEx parseOpts (tmpBackend opts st) request
+  return (fromRaw rawData)
   where parseOpts = generalOptions opts
 
--- Perform cleanup of uploaded files in /tmp.
-cleanup :: (MultipartData, InternalState) -> IO ()
-cleanup (multipartData, internalState) = do
-  removeFileKeys <- forM (files multipartData) $ \file ->
-    runInternalState
-      (register $ do
-          exists <- doesFileExist (fdFilePath file)
-          when exists $ removeFile (fdFilePath file)
-      )
-      internalState
-  mapM_ release removeFileKeys
-  closeInternalState internalState
-
 -- Add multipart extraction support to a Delayed.
 addMultipartHandling :: FromMultipart multipart
                      => MultipartOptions
                      -> Delayed env (multipart -> a)
                      -> Delayed env a
-addMultipartHandling opts Delayed{..} =
-  Delayed { bodyD     = withRequest $ \request -> do
-              fuzzyMultipartCTCheck (contentTypeH request)
-              b <- bodyD
-              b' <- check opts
-              addCleanup (cleanup b')
-              return (b, b')
-          , serverD   = \cs a (b, (multipartData, _st)) req ->
-              case fromMultipart multipartData of
-                Nothing -> FailFatal $
-                  err400 { errBody = "fromMultipart returned Nothing" }
-                Just x  -> fmap ($ x) $
-                  serverD cs a b req
-          , ..
-          }
-
-  where contentTypeH req = fromMaybe "application/octed-stream" $
+addMultipartHandling opts subserver =
+  addBodyCheck subserver contentCheck bodyCheck
+  where
+    contentCheck = withRequest $ \request ->
+      fuzzyMultipartCTCheck (contentTypeH request)
+
+    bodyCheck () = do
+      mpd <- check opts :: DelayedIO MultipartData
+      case fromMultipart mpd of
+        Nothing -> liftRouteResult $ FailFatal
+          err400 { errBody = "fromMultipart returned Nothing" }
+        Just x -> return x
+
+    contentTypeH req = fromMaybe "application/octed-stream" $
```

Comment:
Hmm, that should be `octet-stream` (the typo was there already though). Could you nonetheless quickly push a patch that fixes it as part of this PR? While we're at it.